add "exludeFields" option on findAll, unit tests.

Sean-McQ opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes #9074
<!-- Define findings related to the feature or bug issue. -->
urlOptions is where we define query params, it now accepts an array of field paths for excludingFields

### Occurred changes and/or fixed issues
<!-- Include information of the changes, including collateral areas which have been affected by this PR as requirement or for convenience. -->
added handler logic for "excludedFields" to the function in steve for urlOptions so that an array of strings representing fields (as per the steve documentation for "exclude") will be appended to the url query params being generated.

additional logic was added to the "findAll" action under dashboard-store so that the "excludeFields" option is passed into the "request" action but will also append the "metadata.managedFields" field to all "findAll" actions. In addition, if fields other than "metadata.managedFields" are in "excludeFields" then the loading actions are appended with a boolean for "partial" which isn't used at the moment but can be used to later extend resources so that we know when we don't have a full resource in the collection.

Tests have been added where necessary but an exhaustive test suite was not added for "findAll" due to the sheer size of the function.

### Technical notes summary
<!-- Outline technical changes which may pass unobserved or may help to understand the process of solving the issue -->
See above for the most part. Note: list requests to norman still append "exclude" to the query params but this is ignored by the API and doesn't cause any issues.

### Areas or cases that should be tested
<!-- Areas that should be tested can include Airgap checks, Rancher upgrades, K8s upgrade, etc. -->
check network list request in dev tools to ensure that those requests include the "exclude=metadata.managedFields" query param. Inspect the payloads of the responses to ensure that the requests to steve don't include the "managedFields" field under "metadata"
<!-- Which browser did you use for local testing? The reviewer should test with a different browser. -->
chrome
<!-- Add missing steps or rewrite them if have been missed or to complement existing information. This should define a clear way to reproduce it and not an approximation. -->

### Areas which could experience regressions
<!-- Create a detailed list of areas to be analyzed which may be affected by the changes, which would require a prior research to avoid regressions. -->

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

richard-cox reviewed (changes requested):
- It looks like there was a bad merge at some point, `shell/plugins/steve/urloptions.js` can be removed. If this is needed for unit tests would suggest copying in urlOptions from `shell/plugins/steve/getters.js` ... and using that function in the getter
- I've given the code another pass in terms of where the field is used, and we're still looking good. We should provide QA with a list of locations / features that are affected though, they may pick up on things we've both missed

Comments only
- We don't exclude the field on the two other methods used to fetch resources - `find` and `findMatching`. It's not a big issue, we don't fetch in many cases, and also reduces the impact of the change a little (we're more likely to used managedFields from the context of an individual resource)

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+import _actions from '@shell/plugins/dashboard-store/actions';
+
```
We have been always using TS for the tests, could we stick to it also for this case?
If it's too complicated, could you please list the difficulties you crossed so we can tackle them down another time?

Sean-McQ replied:
There was a type exception showing on my linter on assertions that seemed to be the result of a conflict with the assertions from cypress. Upon further investigation, it looks like we see that conflict in all existing tests as well and we're okay with that so I changed my tests over to TS.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+  };
+  const dispatch = (action, ...args) => {
+    dispatchCalls[action] = [...(dispatchCalls[action] || []), args];
+    if (action === 'request') {
```
No logic in tests. I am actually not sure why we do not have a linting rule here, but to add a condition in a test makes it worthless. Tests have to be deterministic and driven by data, mocks and stubs.

Sean-McQ replied:
removed the logic from the test, instead using the internal logic from jest's mocks to assert calls and params.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+    commitCalls = {};
+  };
+
+  describe('dashboard-store > actions > findAll', () => {
```
Not to be pedant, but the composed description of the test turns into something like 
```
dashboard-store: actions dashboard-store > actions > findAll expects findAll to return a promise
```
It seems kinda redundant the term `action` and `dashboard-store`. I think we can change the format in something more readable, e.g.:
```ts
describe('Store: dashboard', () => {
  // code
  describe('when findAll is dispatched', () => {
    // code
    it('should return a promise', () => {
      // code
    })
  })
})
```

Sean-McQ replied:
Good point, I changed these out to hopefully be more readable.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+  // we're dumping our commits into a map so we can reference them
+  // actions and mutations are stored against their name in an array so we can capture if they're called multiple times
+  // arguments past the name of the action or mutation are stored in an array in case there are multiple
+  const commit = (mutation, ...args) => {
```
All these properties must be defined in a `beforeEach`, letting them be instantiated each time and avoiding internal changes.

PS: If you have seen me doing it somewhere, it was my mistake as well 😅

Sean-McQ replied:
this would create it's own problems, partially because jest's mocks are globals and the calls to these methods are made asynchronously but it would also cause potential performance issues as this function is going to require quite a few tests and several assertions should be made for each combination of parameters provided to it. Jest's documentation prefers these mocks to be defined reset outside of a beforeEach.

https://github.com/jest-community/eslint-plugin-jest/blob/v24.4.0/docs/rules/no-hooks.md

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+  };
+
+  // setting afterEach here seems reasonable but isn't allowed by the linter
+  const resetCalls = () => {
```
This is the outcome of not using the `beforeEach` as assignment part.

Sean-McQ replied:
this is actually the result of our jest unit test flagging the use of beforeEach and afterEach for setup and teardown of tests. I did go ahead and change this to hopefully more closely align with the suggested implementation in the Jest documentation but beforeEach isn't recommended in this case as it would require me to fully call "findAll" between which assertion for which we make quite a few and will probably need even more before we've fully covered it.

https://github.com/jest-community/eslint-plugin-jest/blob/v24.4.0/docs/rules/no-hooks.md

cnotv replied:
As I tried to explain, that rule suggest to avoid both, the hook and external definitions. The second case is worse because it can be mutated and not be deterministic.
If you are not fine with the rule, you can always disable for the file or you can propose to remove it entirely.

Said that, any setup is more comfortable for your debugging is welcome 👍

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+    }
+  };
+  const state = { config: { namespace: 'unitTest' } };
+  const getters = {
```
We should probably write a TODO with refactoring if a single generic action requires all this mocking.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+  };
+
+  describe('dashboard-store > actions > findAll', () => {
+    let findAllPromise;
```
We should not define a variable outside of a test if it has no shared purpose. If that was the case though, it should be assigned in a `beforeEach` block. As it seems to don't be the case, I would rather remove external definitions.

Also, it's usually also preferred to do not define it outside and rather state it every time due readability, but we may have some exceptions in our code given the reasons.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+    });
+
+    it('expects basic findAll to return an array asynchronously', async() => {
+      const findAllReturnValue = await findAllPromise;
```
Where is `findAllPromise` supposed to come from? If this has to be defined outside, we should create a `describe` wrapper where we want to collect all the cases to be analyzed in separated tests as you did.

Sean-McQ replied:
findAllPromise is defined just a few lines above this one and is the un-awaited return value from findAll which is an async function and thus returns a promise. There is a test for that as well.

This has since been updated but the general behavior of the test is the same, we need to test for both the synchronous return value of the async function (which should be a promise) and the async return value after the await keyword in order to ensure that our tests are deterministic and point to the correct type of failure.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+      const findAllReturnValue = await findAllPromise;
+
+      expect(findAllReturnValue).toHaveLength(1);
+      expect(findAllReturnValue[0]).toBe('type:foo');
```
I guess the linter should complain to be `toStrictlyEqual()`

Sean-McQ replied:
`toBe` seemed to be what we were using elsewhere for similar comparisons so that's what I used here.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+      expect(findAllReturnValue[0]).toBe('type:foo');
+    });
+    it('expects basic findAll to dispatch the "request" action once with a url, streaming, and "metadata.managedFields" excluded', () => {
+      expect(dispatchCalls.request).toMatchObject([[{
```
It starts to be very difficult at this state to find out where `dispatchCalls` come from.

cnotv commented on shell/plugins/dashboard-store/__tests__/actions.spec.js:
```diff
@@ -0,0 +1,125 @@
+      }]]);
+    });
+    it('expects findAll with excludedFields to dispatch the "watch" action once', () => {
+      expect(dispatchCalls.watch).toMatchObject([[{ type: 'type:foo', force: false }]]);
```
I seriously struggle to understand anything about this test at this state.
It's usually convenient to have AAA pattern for tests, so the test will define clearly the initial data situation, the action called to return the final assertion.

In this case with one line I seriously don't get what are we supposed to test here.

Sean-McQ replied:
this implementation is almost identical to the way that jest's mock function operate, nonetheless, it's been refactored to use jest's mock function which will similarly return values for when the dispatch function is called, how many times, and with which parameters.

cnotv commented on shell/plugins/steve/__tests__/getters.spec.js:
```diff
@@ -0,0 +1,40 @@
+  describe('dashboard-store > getters > urlFor', () => {
+    // we're not testing function output based off of state or getter inputs here since they are dependencies
+    const state = { config: { baseUrl: 'protocol' } };
+    const getters = {
```
Similar consideration done for the actions are worth also for this case. Assignment done outside of a `beforeEach` lead to issue with mutated objects.

Sean-McQ replied:
https://github.com/jest-community/eslint-plugin-jest/blob/v24.4.0/docs/rules/no-hooks.md

cnotv reviewed (changes requested):
I would like before anything thank you for taking care of something which may be overwhelming. 
There's some considerations to be done, more over about the use of JS over TS and test structure.
Please feel free to address the reasoning or reach out if you want to tackle down together some complexities.

Sean-McQ commented on cypress/e2e/po/pages/page.po.ts:
```diff
@@ -21,7 +21,7 @@ export default class PagePo extends ComponentPo {
   ]) {
     getUrls.forEach((cUrl, i) => {
-      cy.intercept('GET', cUrl).as(`getUrl${ i }`);
+      cy.intercept('GET', `${ cUrl }*`).as(`getUrl${ i }`);
```
updated to make each call in each e2e test specific

Sean-McQ commented on shell/plugins/steve/urloptions.js:
```diff
@@ -29,6 +40,7 @@ export default function urlOptions(url, opt) {
   }
   // End: Limit
 
+  // ToDo: Steve's sort options work differently nowadays (https://github.com/rancher/steve#sort)
```
This ToDo is intentional as updating sort in urlOptions for the current steve spec is out of scope for this issue and would probably require a second urlOptions specific to Norman anyways which would not be a trivial architectural change. Nonetheless, this comment should become useful to any developer running into related issues.

richard-cox replied:
There should be no `todo`'s in code. If there's tech debt there should be an gh issue that's linked in a comment

Sean-McQ replied:
Added comment with https://github.com/rancher/dashboard/issues/9341

I wasn't aware this was a standard as there are currently 44 other ToDos in the codebase currently.

Sean-McQ commented on shell/plugins/dashboard-store/actions.js:
```diff
@@ -184,6 +186,13 @@ export default {
     console.log(`Find All: [${ ctx.state.config.namespace }] ${ type }`); // eslint-disable-line no-console
     opt = opt || {};
+    // we should always exclude managed fields when retreiving resources to cut down on payload size and store memory
+    if (isArray(opt?.excludeFields)) {
```
changed

Sean-McQ commented on shell/plugins/dashboard-store/actions.js:
```diff
@@ -184,6 +186,13 @@ export default {
+    // we should always exclude managed fields when retreiving resources to cut down on payload size and store memory
+    if (isArray(opt?.excludeFields)) {
+      partial = true;
+      opt.excludeFields = [...opt.excludeFields, 'metadata.managedFields'];
```
I went back and forth on this actually, but the spec on this is that we should always exclude metadata.managedFields. Moreover, since we don't use managedFields anywhere, there really isn't a case where we wouldn't want to exclude it and making the excludeFields option override it would simply result in more boilerplate as every developer providing an "excludeFields" array would need to remember to include it.

richard-cox replied:
The spec is we set it by default, not that it's always there. It seems odd to let the user specify their own but not use what they've request, which could be specifically to include managedFields. 

You're right though, we don't have a use case for setting excludes fields. whatever we hit first should be used to drive a change here

Sean-McQ commented on shell/plugins/dashboard-store/actions.js:
```diff
@@ -236,7 +245,9 @@ export default {
           queue = [];
-          commit('loadMulti', { ctx, data: tmp });
+          commit('loadMulti', {
+            ctx, data: tmp, partial
```
It's not currently needed but given the current codebase and our past history of implementing similar or identical options in complicated actions like this one (depaginate and incremental loading are essentially the same thing), a leading indicator of next steps for handling potential issues related to further development of the excludedFields functionality on this action seems appropriate. I added a comment to the code to further clarify.

richard-cox replied:
Can we add them to the mutation functions, so it's visible that side as well?

Sean-McQ replied:
The comments or additional code that needs to be tested? I'd rather not go down the path of fully implementing partial loading as a feature in this issue.

richard-cox replied:
The request is just to add the `partial` property to the loadMulti functions. This makes it clear that in some cases a `partial` property is supplied and avoids any issues if someone wanted to add a partial property for another use case (which i think was the reason why the partial was added when loadMulti is called)

Sean-McQ replied:
Added the partial flag to the load mutations.

Sean-McQ commented on shell/plugins/steve/getters.js:
```diff
@@ -54,6 +54,16 @@ export default {
+    // Exclude
+    // excludeFields should be an array of strings representing the paths of the fields to exclude
+    // only works on Steve but is ignored without error by Norman
+    if (opt.excludeFields && opt.excludeFields.length > 0) {
```
changed

Sean-McQ commented on shell/plugins/steve/getters.js:
```diff
@@ -85,7 +95,7 @@ export default {
     // - a string representing a single namespace - add restriction to the url
     // - an array of namespaces or projects - add restriction as a param
-    if (opt.namespaced && !pAndNFiltering.isApplicable(opt)) {
+    if (opt?.namespaced && !pAndNFiltering.isApplicable(opt)) {
```
Since a default object is not provided in the function, it's possible to call this function without providing the parameter and doing so would cause this function to crash. This implementation is less brittle.

Sean-McQ replied:
There is no other half of this if block, moreover, if opt were undefined then the previous code would crash as javascript would not be able to read the property `namespaced` of undefined so it wouldn't even get to the other half of the if block.

Sean-McQ commented on shell/plugins/steve/urloptions.js:
```diff
@@ -4,6 +4,7 @@ export default function urlOptions(url, opt) {
   opt = opt || {};
 
   // Filter
+  // ToDo: Steve's filter options work differently nowadays (https://github.com/rancher/steve#filter)
```
This ToDo is intentional as updating sort in urlOptions for the current steve spec is out of scope for this issue and would probably require a second urlOptions specific to Norman anyways which would not be a trivial architectural change. Nonetheless, this comment should become useful to any developer running into related issues.

richard-cox replied:
same as https://github.com/rancher/dashboard/pull/9300#discussion_r1265352969

Sean-McQ replied:
see reply in https://github.com/rancher/dashboard/pull/9300#discussion_r1265388301

richard-cox commented on shell/plugins/dashboard-store/actions.js:
```diff
@@ -184,6 +186,13 @@ export default {
     console.log(`Find All: [${ ctx.state.config.namespace }] ${ type }`); // eslint-disable-line no-console
     opt = opt || {};
+    // we should always exclude managed fields when retreiving resources to cut down on payload size and store memory
+    if (isArray(opt?.excludeFields)) {
```
There's two parts to the comment

1) Steve specific code should be in the steve store and not dashboard store (still pending)
  - The boundary is the files. steve actions inherit and in places override generic dashboard actions
2) The steve store caters for both steve, norman and kube api's
  - The boundary doesn't currently exist. 
  - Checking for v1 might work, but it would need to be from the start of the string (kube api contains v1)

Sean-McQ replied:
Given that other APIs don't seem to error out when passing in this parameter even if it's not used, I think we're safe with a simple '/v1/' check on the string unless you've got a simple bit of code you think would work well here.

The right solution would be to refactor the code so that the steve plugin (there isn't actually a steve store) is only used for code that pertains to the steve API but I would say that falls outside the scope of this issue.

Sean-McQ replied:
updated

richard-cox reviewed (changes requested):
See comments in previous review conversations

Sean-McQ commented on cypress/e2e/po/pages/repositories.po.ts:
```diff
@@ -34,7 +34,7 @@ export default class ReposListPagePo extends PagePo {
 
   waitForGoTo(endpoint: string) {
-    const interceptName = `validateGoTo${ Date.now() }`;
+    const interceptName = `validateGoTo${ Date.now() }*`;
```
updated

Sean-McQ commented:
`find` and `findMatching` don't have matching excludeField logic since these two are for finding specific resources via the API, the reason why we would want to exclude fields from an API request boil down to decreasing payload size (and thus memory client-side) and/or server load serializing large responses. The benefit for both of these is minimal for targeted requests like what we would be using find and findMatching for and doing so creates additional code complexity with little benefit. A good example would be that we don't implement incremental loading on findMatching because the benefit would be minimal.

This addresses all comments on conversations. Please let me know if you've any further feedback.

richard-cox reviewed (changes requested):
Three were not addressed in the last round
- https://github.com/rancher/dashboard/pull/9300#discussion_r1265362454
- https://github.com/rancher/dashboard/pull/9300#discussion_r1265384768
- https://github.com/rancher/dashboard/pull/9300#discussion_r1265386851

Sean-McQ commented:
addressed the three above.
